### PR TITLE
 update escrow contract

### LINKED ISSUE #275: Contract changes to refine trusted handlers & gas optimization

### PR DESCRIPTION
**Pull request by leric7**
## Description

`Escrow` smart contract needs to be updated in terms of trusted handler & its own state management.

## Summary of changes

- Refactored `Escrow` contract
   - Removed `trustedHandlers`, and added `exchangeOracle` as one of the known trusted handler.
   - Updated escrow status enum, so that it can be self explanatory.
   - Removed `setup` function, and merged it into the constructor.
   - Updated escrow functions, so that it can be only called in the correct state, and from the correct handler.
- Refactored `EscrowFactory` contract.
   - Updated `createEscrow` function, so that it can create/setup/allocate new escrow.
- Updated `Staking` contract.
   - Added a new function `allocateFrom`, so that `EscrowFactory` can directly allocate funds from job launcher to the escrow.

## Related issues
[Keywords for linking issues](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)

Closes #275 

## Operational checklist

- [ ] All new functionality is covered by tests
- [ ] Any related documentation has been changed or added

### CODE REVIEW

File: packages/core/contracts/Escrow.sol [hunk 1]
```diff
@@ -55,236 +64,175 @@ contract Escrow is IEscrow, ReentrancyGuard {
...
+
+    function complete() external override notExpired {
+        require(_msgSender() == reputationOracle, 'Invalid handler');
+        require(status == EscrowStatuses.Paid, 'Escrow not in Paid state');
+
+        status = EscrowStatuses.Completed;
+        emit Completed();
+    }
+
+    function abort() external override trusted notComplete notPaid {
```
Comment (portuu3) [inline packages/core/contracts/Escrow.sol:206]: why abort and cancel? can we combine them?

File: packages/core/contracts/Escrow.sol [hunk 2]
```diff
@@ -322,33 +270,26 @@ contract Escrow is IEscrow, ReentrancyGuard {
...
+            _msgSender() == launcher ||
+                _msgSender() == exchangeOracle ||
+                _msgSender() == reputationOracle ||
+                _msgSender() == recordingOracle,
+            'Not trusted handler'
+        );
         _;
     }
 
     modifier notComplete() {
```
Comment (portuu3) [inline packages/core/contracts/Escrow.sol:283]: if we combine abort and cancel this is not necessary

File: packages/core/contracts/Escrow.sol [hunk 3]
```diff
@@ -322,33 +270,26 @@ contract Escrow is IEscrow, ReentrancyGuard {
...
         require(
-            status != EscrowStatuses.Complete,
-            'Escrow in Complete status state'
+            status != EscrowStatuses.Completed,
+            'Escrow in Completed state'
         );
         _;
     }
 
     modifier notPaid() {
```
Comment (portuu3) [inline packages/core/contracts/Escrow.sol:291]: if we combine abort and cancel this is not necessary

File: packages/core/contracts/EscrowFactory.sol [hunk 4]
```diff
@@ -34,27 +34,57 @@ contract EscrowFactory is OwnableUpgradeable, UUPSUpgradeable {
 
     function createEscrow(
```
Comment (portuu3) [inline packages/core/contracts/EscrowFactory.sol:35]: fund the escrow is missing

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/core/contracts/Escrow.sol [hunk 5]
```diff
@@ -55,236 +64,175 @@ contract Escrow is IEscrow, ReentrancyGuard {
...
-        status = EscrowStatuses.Complete;
-        emit Completed();
+
+        status = EscrowStatuses.Exchanged;
+        emit Exchanged();
     }
 
     function storeResults(
-        address _sender,
+        address _worker,
```
[the remark this answers is not in the document]
Comment (eugenvoronov) [reply packages/core/contracts/Escrow.sol:122]: Can we modify this method, depends on status we need to add new logic:
- `rawResultsUrl` to store a url to s3 bucket with raw annotations from Exchange Oracle
- `finalResultsUrl ` to store a url to s3 bucket with raw annotations from Recording Oracle

File: packages/core/contracts/EscrowFactory.sol (hunk 4 above)
Comment (leric7) [reply packages/core/contracts/EscrowFactory.sol:35]: Do we want to fund the escrow by job launcher? I think escrow can be funded from job requesters after the escrow is created?

**Review by portuu3 — CHANGES_REQUESTED** [review]

**Comment by posix4e** [comment]
I don't understand why we want to remove trusted handlers or do any other status updates in the same commit. Can we break out staking.

**Comment by leric7** [comment]
We won't do these changes

**Pull request closed without merging**